File: server/utils/cq-parser.js
/**
 * CQ码解析器
 * 用于解析和构建CQ码消息段
 * 
 * @module cq-parser
 * @version 1.0.0
 * @author KiBot Team
 */

/**
 * CQ码解析器
 */
export class CQParser {
  /**
   * 解析CQ码字符串为消息段数组
   * @param {string} message - 包含CQ码的消息字符串
   * @returns {Array} 消息段数组
   */
  static parse(message) {
    if (!message || typeof message !== 'string') {
      return [];
    }

    const segments = [];
    const cqRegex = /\[CQ:([^,\]]+)((?:,[^\]]+)?)\]/g;
    let lastIndex = 0;
    let match;

    while ((match = cqRegex.exec(message)) !== null) {
      // 添加CQ码之前的文本
      if (match.index > lastIndex) {
        const text = message.substring(lastIndex, match.index);
        if (text) {
          segments.push({
            type: 'text',
            data: { text }
          });
        }
      }

      // 解析CQ码
      const type = match[1];
      const paramsStr = match[2] || '';
      const data = this.parseParams(paramsStr);

      segments.push({
        type,
        data
      });

      lastIndex = match.index + match[0].length;
    }

    // 添加最后剩余的文本
    if (lastIndex < message.length) {
      const text = message.substring(lastIndex);
      if (text) {
        segments.push({
          type: 'text',
          data: { text }
        });
      }
    }

    return segments;
  }

  /**
   * 解析CQ码参数
   * @param {string} paramsStr - 参数字符串
   * @returns {Object} 参数对象
   */
  static parseParams(paramsStr) {
    const params = {};
    if (!paramsStr) return params;

    // 移除开头的逗号
    const cleanStr = paramsStr.startsWith(',') ? paramsStr.substring(1) : paramsStr;
    
    // 使用正则匹配 key=value 对
    const paramRegex = /([^,=]+)=([^,]*?)(?=,|$)/g;
    let match;

    while ((match = paramRegex.exec(cleanStr)) !== null) {
      const key = match[1].trim();
      let value = match[2].trim();
      
      // 解码特殊字符
      value = this.decodeCQValue(value);
      
      params[key] = value;
    }

    return params;
  }

  /**
   * 解码CQ码值中的特殊字符
   * @param {string} value - 待解码的值
   * @returns {string} 解码后的值
   */
  static decodeCQValue(value) {
    return value
      .replace(/&amp;/g, '&')
      .replace(/&#91;/g, '[')
      .replace(/&#93;/g, ']')
      .replace(/&#44;/g, ',');
  }

  /**
   * 编码CQ码值中的特殊字符
   * @param {string} value - 待编码的值
   * @returns {string} 编码后的值
   */
  static encodeCQValue(value) {
    return String(value)
      .replace(/&/g, '&amp;')
      .replace(/\[/g, '&#91;')
      .replace(/\]/g, '&#93;')
      .replace(/,/g, '&#44;');
  }

  /**
   * 构建CQ码字符串
   * @param {string} type - CQ码类型
   * @param {Object} data - CQ码数据
   * @returns {string} CQ码字符串
   */
  static build(type, data = {}) {
    const params = Object.entries(data)
      .map(([key, value]) => `${key}=${this.encodeCQValue(value)}`)
      .join(',');
    
    return params ? `[CQ:${type},${params}]` : `[CQ:${type}]`;
  }

  /**
   * 从消息段数组中提取纯文本
   * @param {Array} segments - 消息段数组
   * @returns {string} 纯文本
   */
  static extractText(segments) {
    return segments
      .filter(seg => seg.type === 'text')
      .map(seg => seg.data.text)
      .join('');
  }

  /**
   * 从消息段数组中提取指定类型的段
   * @param {Array} segments - 消息段数组
   * @param {string} type - 消息段类型
   * @returns {Array} 指定类型的消息段数组
   */
  static extractByType(segments, type) {
    return segments.filter(seg => seg.type === type);
  }

  /**
   * 检查消息段中是否包含指定类型
   * @param {Array} segments - 消息段数组
   * @param {string} type - 消息段类型
   * @returns {boolean} 是否包含
   */
  static hasType(segments, type) {
    return segments.some(seg => seg.type === type);
  }

  /**
   * 将消息段数组转换为字符串
   * @param {Array} segments - 消息段数组
   * @returns {string} 消息字符串
   */
  static stringify(segments) {
    return segments.map(seg => {
      if (seg.type === 'text') {
        return seg.data.text || '';
      }
      return this.build(seg.type, seg.data);
    }).join('');
  }
}

/**
 * 常用CQ码构建器
 */
export class CQBuilder {
  /**
   * 构建文本消息
   * @param {string} text - 文本内容
   * @returns {string} 文本
   */
  static text(text) {
    return text;
  }

  /**
   * 构建@某人
   * @param {number|string} qq - QQ号，或"all"表示@全体成员
   * @returns {string} CQ码
   */
  static at(qq) {
    return CQParser.build('at', { qq });
  }

  /**
   * 构建表情
   * @param {number} id - 表情ID
   * @returns {string} CQ码
   */
  static face(id) {
    return CQParser.build('face', { id });
  }

  /**
   * 构建图片
   * @param {string} file - 图片文件名、URL或base64
   * @param {Object} options - 其他选项
   * @returns {string} CQ码
   */
  static image(file, options = {}) {
    return CQParser.build('image', { file, ...options });
  }

  /**
   * 构建语音
   * @param {string} file - 语音文件名或URL
   * @returns {string} CQ码
   */
  static record(file) {
    return CQParser.build('record', { file });
  }

  /**
   * 构建视频
   * @param {string} file - 视频文件名或URL
   * @returns {string} CQ码
   */
  static video(file) {
    return CQParser.build('video', { file });
  }

  /**
   * 构建回复
   * @param {number|string} id - 消息ID
   * @returns {string} CQ码
   */
  static reply(id) {
    return CQParser.build('reply', { id });
  }

  /**
   * 构建戳一戳
   * @param {number|string} qq - QQ号
   * @returns {string} CQ码
   */
  static poke(qq) {
    return CQParser.build('poke', { qq });
  }

  /**
   * 构建链接分享
   * @param {string} url - 链接URL
   * @param {string} title - 标题
   * @param {Object} options - 其他选项
   * @returns {string} CQ码
   */
  static share(url, title, options = {}) {
    return CQParser.build('share', { url, title, ...options });
  }

  /**
   * 构建音乐分享
   * @param {string} type - 音乐类型（qq/163/xm）
   * @param {string} id - 音乐ID
   * @returns {string} CQ码
   */
  static music(type, id) {
    return CQParser.build('music', { type, id });
  }

  /**
   * 构建自定义音乐分享
   * @param {string} url - 跳转URL
   * @param {string} audio - 音频URL
   * @param {string} title - 标题
   * @param {Object} options - 其他选项
   * @returns {string} CQ码
   */
  static customMusic(url, audio, title, options = {}) {
    return CQParser.build('music', { type: 'custom', url, audio, title, ...options });
  }

  /**
   * 构建位置分享
   * @param {number} lat - 纬度
   * @param {number} lon - 经度
   * @param {string} title - 标题
   * @param {Object} options - 其他选项
   * @returns {string} CQ码
   */
  static location(lat, lon, title, options = {}) {
    return CQParser.build('location', { lat, lon, title, ...options });
  }

  /**
   * 构建合并转发消息
   * @param {string} id - 合并转发ID
   * @returns {string} CQ码
   */
  static forward(id) {
    return CQParser.build('forward', { id });
  }

  /**
   * 构建XML消息
   * @param {string} data - XML数据
   * @returns {string} CQ码
   */
  static xml(data) {
    return CQParser.build('xml', { data });
  }

  /**
   * 构建JSON消息
   * @param {string} data - JSON数据
   * @returns {string} CQ码
   */
  static json(data) {
    return CQParser.build('json', { data });
  }
}

/**
 * 消息段辅助类
 */
export class MessageSegment {
  /**
   * 检查是否是图片消息段
   * @param {Object} segment - 消息段
   * @returns {boolean}
   */
  static isImage(segment) {
    return segment && segment.type === 'image';
  }

  /**
   * 检查是否是@消息段
   * @param {Object} segment - 消息段
   * @returns {boolean}
   */
  static isAt(segment) {
    return segment && segment.type === 'at';
  }

  /**
   * 检查是否是文本消息段
   * @param {Object} segment - 消息段
   * @returns {boolean}
   */
  static isText(segment) {
    return segment && segment.type === 'text';
  }

  /**
   * 检查是否是回复消息段
   * @param {Object} segment - 消息段
   * @returns {boolean}
   */
  static isReply(segment) {
    return segment && segment.type === 'reply';
  }

  /**
   * 获取图片URL
   * @param {Object} segment - 图片消息段
   * @returns {string|null}
   */
  static getImageUrl(segment) {
    if (!this.isImage(segment)) return null;
    return segment.data?.url || null;
  }

  /**
   * 获取@的QQ号
   * @param {Object} segment - @消息段
   * @returns {string|null}
   */
  static getAtQQ(segment) {
    if (!this.isAt(segment)) return null;
    return segment.data?.qq || null;
  }

  /**
   * 获取文本内容
   * @param {Object} segment - 文本消息段
   * @returns {string|null}
   */
  static getText(segment) {
    if (!this.isText(segment)) return null;
    return segment.data?.text || null;
  }

  /**
   * 获取回复的消息ID
   * @param {Object} segment - 回复消息段
   * @returns {string|null}
   */
  static getReplyId(segment) {
    if (!this.isReply(segment)) return null;
    return segment.data?.id || null;
  }
}

export default {
  CQParser,
  CQBuilder,
  MessageSegment
};

File: server/core/python-plugin-system/cq_parser.py
"""
CQ码解析和构建工具
支持QQ消息中的富媒体内容处理
"""

import re
from typing import List, Dict, Any, Optional
from urllib.parse import quote, unquote


class CQParser:
    """CQ码解析器"""
    
    # CQ码正则表达式
    CQ_CODE_PATTERN = re.compile(r'\[CQ:([a-zA-Z]+)((?:,[a-zA-Z]+=[^,\]]*)*)\]')
    
    @classmethod
    def parse(cls, message: str) -> List[Dict[str, Any]]:
        """
        解析消息为消息段数组
        
        Args:
            message: 原始消息字符串
        
        Returns:
            消息段列表
        """
        segments = []
        last_end = 0
        
        for match in cls.CQ_CODE_PATTERN.finditer(message):
            # 添加纯文本段
            if match.start() > last_end:
                text = message[last_end:match.start()]
                if text:
                    segments.append({
                        'type': 'text',
                        'data': {'text': text}
                    })
            
            # 解析CQ码
            cq_type = match.group(1)
            params_str = match.group(2)
            
            # 解析参数
            data = {}
            if params_str:
                for param in params_str.split(','):
                    if not param:
                        continue
                    if '=' in param:
                        key, value = param.split('=', 1)
                        data[key] = cls._unescape(value)
            
            segments.append({
                'type': cq_type,
                'data': data
            })
            
            last_end = match.end()
        
        # 添加剩余文本
        if last_end < len(message):
            text = message[last_end:]
            if text:
                segments.append({
                    'type': 'text',
                    'data': {'text': text}
                })
        
        return segments
    
    @classmethod
    def extract_text(cls, segments: List[Dict[str, Any]]) -> str:
        """提取纯文本内容"""
        texts = []
        for segment in segments:
            if segment['type'] == 'text':
                texts.append(segment['data'].get('text', ''))
        return ''.join(texts)
    
    @classmethod
    def extract_by_type(cls, segments: List[Dict[str, Any]], seg_type: str) -> List[Dict[str, Any]]:
        """提取指定类型的消息段"""
        return [seg for seg in segments if seg['type'] == seg_type]
    
    @classmethod
    def has_type(cls, segments: List[Dict[str, Any]], seg_type: str) -> bool:
        """检查是否包含指定类型的消息段"""
        return any(seg['type'] == seg_type for seg in segments)
    
    @staticmethod
    def _unescape(text: str) -> str:
        """反转义CQ码参数"""
        return text.replace('&amp;', '&').replace('&#91;', '[').replace('&#93;', ']').replace('&#44;', ',')
    
    @staticmethod
    def _escape(text: str) -> str:
        """转义CQ码参数"""
        return text.replace('&', '&amp;').replace('[', '&#91;').replace(']', '&#93;').replace(',', '&#44;')


class CQBuilder:
    """CQ码构建器"""
    
    @staticmethod
    def _build(cq_type: str, **params) -> str:
        """构建CQ码"""
        if not params:
            return f"[CQ:{cq_type}]"
        
        param_str = ','.join(
            f"{k}={CQParser._escape(str(v))}" 
            for k, v in params.items()
        )
        return f"[CQ:{cq_type},{param_str}]"
    
    @classmethod
    def at(cls, qq: int) -> str:
        """@某人"""
        return cls._build('at', qq=qq)
    
    @classmethod
    def at_all(cls) -> str:
        """@全体成员"""
        return cls._build('at', qq='all')
    
    @classmethod
    def face(cls, face_id: int) -> str:
        """QQ表情"""
        return cls._build('face', id=face_id)
    
    @classmethod
    def image(cls, file: str, url: Optional[str] = None, cache: bool = True) -> str:
        """
        图片
        
        Args:
            file: 图片文件名或URL
            url: 图片URL（可选）
            cache: 是否使用缓存
        """
        params = {'file': file}
        if url:
            params['url'] = url
        if not cache:
            params['cache'] = 0
        return cls._build('image', **params)
    
    @classmethod
    def record(cls, file: str, magic: bool = False) -> str:
        """
        语音
        
        Args:
            file: 语音文件名
            magic: 是否变声
        """
        params = {'file': file}
        if magic:
            params['magic'] = 1
        return cls._build('record', **params)
    
    @classmethod
    def video(cls, file: str) -> str:
        """视频"""
        return cls._build('video', file=file)
    
    @classmethod
    def reply(cls, message_id: int) -> str:
        """回复消息"""
        return cls._build('reply', id=message_id)
    
    @classmethod
    def poke(cls, qq: int) -> str:
        """戳一戳"""
        return cls._build('poke', qq=qq)
    
    @classmethod
    def share(cls, url: str, title: str, content: Optional[str] = None, image: Optional[str] = None) -> str:
        """
        分享链接
        
        Args:
            url: 链接URL
            title: 标题
            content: 内容描述
            image: 封面图URL
        """
        params = {'url': url, 'title': title}
        if content:
            params['content'] = content
        if image:
            params['image'] = image
        return cls._build('share', **params)
    
    @classmethod
    def music(cls, music_type: str, music_id: str) -> str:
        """
        音乐分享
        
        Args:
            music_type: 音乐平台 (qq/163/xm)
            music_id: 音乐ID
        """
        return cls._build('music', type=music_type, id=music_id)
    
    @classmethod
    def custom_music(cls, url: str, audio: str, title: str, content: Optional[str] = None, image: Optional[str] = None) -> str:
        """
        自定义音乐分享
        
        Args:
            url: 点击链接
            audio: 音频URL
            title: 标题
            content: 描述
            image: 封面图URL
        """
        params = {
            'type': 'custom',
            'url': url,
            'audio': audio,
            'title': title
        }
        if content:
            params['content'] = content
        if image:
            params['image'] = image
        return cls._build('music', **params)
    
    @classmethod
    def location(cls, lat: float, lon: float, title: Optional[str] = None, content: Optional[str] = None) -> str:
        """
        位置分享
        
        Args:
            lat: 纬度
            lon: 经度
            title: 位置标题
            content: 位置描述
        """
        params = {'lat': lat, 'lon': lon}
        if title:
            params['title'] = title
        if content:
            params['content'] = content
        return cls._build('location', **params)
    
    @classmethod
    def forward(cls, forward_id: str) -> str:
        """合并转发"""
        return cls._build('forward', id=forward_id)
    
    @classmethod
    def json(cls, data: str) -> str:
        """JSON消息"""
        return cls._build('json', data=data)
    
    @classmethod
    def xml(cls, data: str) -> str:
        """XML消息"""
        return cls._build('xml', data=data)

